packages/runtime/container-runtime/src/summarizerHeuristics.ts

```diff
@@ -39,46 +59,70 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
             refSequenceNumber: refSequenceNumber ?? this.lastOpSequenceNumber,
             summaryTime: Date.now(),
         };
+
+        this.numSystemOpsBefore = this.numSystemOps;
+        this.numNonSystemOpsBefore = this.numNonSystemOps;
     }
 
     public markLastAttemptAsSuccessful() {
         this._lastSuccessfulSummary = { ...this.lastAttempt };
+
+        this.numSystemOps -= this.numSystemOpsBefore;
+        this.numSystemOpsBefore = 0;
+
+        this.numNonSystemOps -= this.numNonSystemOpsBefore;
+        this.numNonSystemOpsBefore = 0;
     }
 }
 
 /**
  * This class contains the heuristics for when to summarize.
  */
 export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
-    private readonly idleTimer: Timer;
+    private readonly idleTimer: Timer | undefined;
+    private readonly runSummarize: (reason: SummarizeReason) => void;
 
     public constructor(
         private readonly heuristicData: ISummarizeHeuristicData,
         private readonly configuration: ISummaryConfiguration,
-        private readonly trySummarize: (reason: SummarizeReason) => void,
+        trySummarize: (reason: SummarizeReason) => void,
         private readonly minOpsForAttemptOnClose = 50,
+        private readonly summarizeStrategies: ISummarizeHeuristicStrategy[] = getDefaultSummarizeHeuristicStrategies(),
+        useIdleTimerStrategy: boolean = true,
     ) {
-        this.idleTimer = new Timer(
-            this.configuration.idleTime,
-            () => this.trySummarize("idle"));
+        if (useIdleTimerStrategy) {
+            this.idleTimer = new Timer(
+                this.configuration.idleTime,
+                () => this.runSummarize("idle"));
+        }
+
+        this.runSummarize = (reason: SummarizeReason) => {
+            this.idleTimer?.clear();
+
+            // We shouldn't attempt a summary if there are no new processed ops
+            const opsSinceLastAck = this.opsSinceLastAck;
+            if (opsSinceLastAck > 0) {
+                trySummarize(reason);
+            }
+        };
     }
 
     public get opsSinceLastAck(): number {
         return this.heuristicData.lastOpSequenceNumber - this.heuristicData.lastSuccessfulSummary.refSequenceNumber;
     }
 
+    public start() {
+        this.idleTimer?.start();
```

Comment:
I think it should instead call run() method. Two reasons:
- If we have some ops accumulated already, we should consider to summarize them
- If we have no ops accumulated, we should not set timer (as we should not summarize nothing).

That said, it's correct if we loaded from latest summary.
In such case our op counting would be correct.
I'm not sure how code works today if it's not the case. I think we have no way to accurately collect data about user & system ops, we can only see total ops (as we learn about reference sequence number of a summary too late).
I do not even know if it worked correctly before your changes.
Let's chat about it.